finality-aleph/src/sync/substrate/chain_status.rs

```diff
@@ -0,0 +1,147 @@
+use std::{
+    fmt::{Display, Error as FmtError, Formatter},
+    marker::PhantomData,
+};
+
+use aleph_primitives::{BlockNumber, ALEPH_ENGINE_ID};
+use log::warn;
+use sp_blockchain::Backend;
+use sp_runtime::{
+    generic::BlockId as SubstrateBlockId,
+    traits::{Block as BlockT, Header as SubstrateHeader},
+};
+
+use crate::{
+    justification::backwards_compatible_decode,
+    sync::{substrate::Justification, BlockStatus, ChainStatus, Header, LOG_TARGET},
+    AlephJustification,
+};
+
+/// What can go wrong when checking chain status
+#[derive(Debug)]
+pub enum Error<B: BlockT> {
+    MissingHash(B::Hash),
+    MissingJustification(B::Hash),
+}
+
+impl<B: BlockT> Display for Error<B> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use Error::*;
+        match self {
+            MissingHash(hash) => {
+                write!(f, "no block for hash {:?}", hash)
+            }
+            MissingJustification(hash) => {
+                write!(
+                    f,
+                    "no justification for finalized block with hash {:?}",
+                    hash
+                )
+            }
+        }
+    }
+}
+
+/// Substrate implementation of ChainStatus trait
+struct SubstrateChainStatus<B, BE>
+where
+    BE: Backend<B>,
+    B: BlockT,
+    B::Header: SubstrateHeader<Number = BlockNumber>,
+{
+    client: BE,
+    _phantom: PhantomData<B>,
+}
+
+impl<B, BE> SubstrateChainStatus<B, BE>
+where
+    BE: Backend<B>,
+    B: BlockT,
+    B::Header: SubstrateHeader<Number = BlockNumber>,
+{
+    fn header(&self, hash: B::Hash) -> Option<B::Header> {
+        let id = SubstrateBlockId::<B>::Hash(hash);
+        self.client
+            .header(id)
+            .expect("substrate client must respond")
```

Comment:
idk what is the policy, but maybe it would be better to be defensive here and convert `Err` to `None`? or extend `Error` enum and return `Result` here

Reply:
Yep, we thought about it. The thing is we already use client in many places and use `expect` on the result. To be fair, if client is not working something really bad is going on an we probably can't fix it.

That said, from what I see, substrate does some error handling in this case, so I can add those here.